### PR TITLE
PUBDEV-9712: add model_summary to rulefit models

### PR DESCRIPTION
**Pull request by koniecsveta**
https://h2oai.atlassian.net/browse/PUBDEV-7912

combined model summary information of tree ensemble, rule ensemble and resulting linear model

### CODE REVIEW

File: h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py [hunk 1]
```diff
@@ -17,6 +17,7 @@ def titanic():
     rfit.train(training_frame=train, x=x, y="survived")
 
     print(rfit.rule_importance())
+    print(rfit._model_json["output"]["model_summary"])
```
[the remark this answers is not in the document]
Comment (koniecsveta) [reply h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py]: 31f8d1a

File: h2o-algos/src/main/java/hex/tree/TreeStats.java [hunk 2]
```diff
@@ -38,4 +38,18 @@ public String toString() {
...
+    if (otherTreeStats._min_depth < this._min_depth) this._min_depth = otherTreeStats._min_depth;
+    if (otherTreeStats._max_depth > this._max_depth) this._max_depth = otherTreeStats._max_depth;
+    if (otherTreeStats._min_leaves < this._min_leaves) this._min_leaves = otherTreeStats._min_leaves;
+    if (otherTreeStats._max_leaves > this._max_leaves) this._max_leaves = otherTreeStats._max_leaves;
+    this._byte_size += otherTreeStats._byte_size;
+    this._num_trees += otherTreeStats._num_trees;
+    this._sum_depth += otherTreeStats._sum_depth;
+    this._sum_leaves += otherTreeStats._sum_leaves;
+    this._num_trees += otherTreeStats._num_trees;
+    _mean_depth = ((float) _sum_depth / _num_trees);
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/tree/TreeStats.java:52]: this code is copy-pasted from the `updateBy` method, I would rather put it into a `private void updateMeans()` also a unit test for this method would be nice

File: h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 
 def titanic():
-    df = h2o.import_file("https://s3.amazonaws.com/h2o-public-test-data/smalldata/gbm_test/titanic.csv",
+    df = h2o.import_file(pyunit_utils.locate("smalldata/gbm_test/titanic.csv"),
```
Comment (maurever) [inline h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py:9]: :+1:

File: h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py [hunk 4]
```diff
@@ -17,6 +17,8 @@ def titanic():
     rfit.train(training_frame=train, x=x, y="survived")
 
     print(rfit.rule_importance())
+    assert rfit._model_json["output"]["model_summary"] is not None
```
Comment (maurever) [inline h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py:20]: Assert with a message would be better. :)

**Review by maurever — COMMENTED** [review]
Looks good, just a few suggestions.

File: h2o-algos/src/main/java/hex/tree/TreeStats.java (hunk 2 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/tree/TreeStats.java]: fac8196

**Review by honzasterba — APPROVED** [review]

File: h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py (hunk 4 above)
Comment (koniecsveta) [reply h2o-py/tests/testdir_algos/rulefit/pyunit_titanic_rulefit.py]: 49d3cc2

**Review by maurever — APPROVED** [review]

**Comment by ledell** [comment]
@koniecsveta Can you paste an example of what the output looks like? I assume it's the same in R and Python?  Thanks!

**Comment by koniecsveta** [comment]
> @koniecsveta Can you paste an example of what the output looks like? I assume it's the same in R and Python? Thanks!

yes, it's the same: output is in tabular form, table is created in java backend and then passed to clients.

example in Python:
Screenshot 2020-12-02 at 10 57 36

example in R:
Screenshot 2020-12-02 at 11 00 13

**Pull request merged**